DSEGOG-3 DSEGOG-5 DSEGOG-6 Adding initial basic table and column functionality

sam-glendenning opened the pull request:
## Description
This includes the use of React Table for the headless table functionality and a small amount of Material UI to make it look somewhat presentable. Right now this focuses on three main pieces of functionality:

1. Display of data in a table (data itself is randomly generated locally for now while we don't have an API)
2. Table pagination
3. Table column sorting (currently columns are multi-sorted but this will likely be changed to be single sort as multi-sort is not a specific OG requirement)

## Testing instructions
Double check that you are set up with yarn 2 and that it can zero-install all the required packages with `yarn install`. This should be an automatic process thanks to the `.yarnrc.yml` file. You should also be able to open the project correctly in VS Code (there is a configuration for this in `.yarn/sdks`.
Then try out the table in the browser (data display and pagination should work fine and sorting a column will result in the column header showing as sorted but not actually manipulating the table). Also make sure adding and removing columns by ticking and unticking the checkboxes works as expected. The checkboxes should refresh with what data is available based on what page you're on.

## Agile board tracking
Closes [DSEGOG-3](https://stfc.atlassian.net/browse/DSEGOG-3), [DSEGOG-5](https://stfc.atlassian.net/browse/DSEGOG-5), [DSEGOG-6](https://stfc.atlassian.net/browse/DSEGOG-6)

Commit:
Adding basic RecordTable component

Contains a basic record structure, randomly generated local data, useQueries for fetching the data and its count and a basic paginated useTable instance

Commit:
Replacing HTML table with MUI Table elements

Commit:
Adding MUI zero-based pagination

Commit:
Defining basic query parameters and filters

Commit:
Adding server-side column sorting

Works with multiple columns, since the method mimics DataGateway's functionality. Can always amend later to a single column if needed.

Also changing to single quotes to ensure better compatibility down the line.

Commit:
Adding basics for pushing and popping columns onto table

The data schema has now been extended to factor in generic channels into record entries. For every page of data that is currently available, all of the available columns that can be displayed can be manually selected to display on the table by ticking its checkbox. This appends it to the table on the right.

The code is a complete mess right now so this will need to be refactored before continuing, otherwise it's just going to be impossible to maintain

Commit:
Checked checkboxes do not carry over onto other pages anymore

Previously, incorrect behaviour meant that checking a checkbox would mean the checkbox in the same position on other pages would also be checked (although this wouldn't cause its column to appear/disappear). This has now been fixed so that a checkbox is checked based on what columns are currently being displayed

Commit:
Adding unit tests for table and columnCheckboxes

Uses React Testing Library and snapshots

louise-davies committed:
Committing some yarn 2 stuff & vscode yarn config

Commit:
DSEGOG-3 DSEGOG-5 DSEGOG-6 Adding unit tests for dataHeader

Commit:
DSEGOG-3 DSEGOG-5 DSEGOG-6 Adding further unit tests and refactoring where necessary

Commit:
Adding clarification on singleSpaReact call

Commit:
Attempting CI build workflow addition

This likely won't work, I just want to see where I'm at with it

Commit:
Fixing spelling mistake in workflow action

Commit:
Fixing linting error

Commit:
Adding codecov.yml

louise-davies commented on src/index.tsx:
```diff
@@ -28,17 +28,22 @@ function domElementGetter(): HTMLElement {
   React,
   ReactDOM,
-  rootComponent: App,
+  // rootComponent: App,
   domElementGetter,
+  loadRootComponent: () =>
+    new Promise((resolve, reject) => resolve(() => <App />)),
```
I don't see why we need to use `loadRootComponent` here - and I think it would be good as well to move the error boundary stuff from `App` to here

sam-glendenning replied:
@louise-davies Made the changes

Commit:
DSEGOG-3 Making App a functional component and moving singleSpa error catching to errorBoundary opt

This fixes any warnings/errors thrown during unit tests. It also simplifies the error handling process and cleans up the code for App.tsx

louise-davies reviewed (approved):
Looks good - I've not gone over all the WIP code with a fine tooth comb as I know most of it is subject to change but it looks fine and works well so far :)
